packages/seacas/libraries/ioss/src/Ioss_DynamicTopology.C

```diff
@@ -36,670 +37,882 @@
+    locate_state_impl(db, targetTime, compare, loc);
+  }
+}
+
+void locate_db_state_impl(Ioss::DatabaseIO* db, double targetTime, DatabaseState& loc)
+{
+  auto region = db->get_region();
+  auto changeSet = Ioss::ChangeSetFactory::create(region);
+  changeSet->populate_change_sets();
+
+  for(unsigned csIndex=0; csIndex<changeSet->size(); csIndex++) {
+    auto csdb = changeSet->open_change_set(csIndex, Ioss::QUERY_TIMESTEPS_ONLY);
+    locate_state(csdb, targetTime, loc);
+    changeSet->close_change_set(csIndex);
+  }
+}
+
+void get_db_time_impl(Ioss::DatabaseIO* db, double init_time,
+                      StateLocatorCompare comparator,
+                      DatabaseState& loc)
+{
+  auto region = db->get_region();
+  auto changeSet = Ioss::ChangeSetFactory::create(region);
+  changeSet->populate_change_sets();
+
+  double best_time = init_time;
+
+  for(unsigned csIndex=0; csIndex<changeSet->size(); csIndex++) {
+    auto csdb = changeSet->open_change_set(csIndex, Ioss::QUERY_TIMESTEPS_ONLY);
+
+    std::vector<double> timesteps = csdb->get_db_step_times();
+    int stepCount = static_cast<int>(timesteps.size());
+
+    for (int i = 1; i <= stepCount; i++) {
+      if (comparator(timesteps[i-1], best_time)) {
+        loc.time  = timesteps[i-1];
+        loc.state = i;
+        loc.changeSet = changeSet->get_change_set_name(csIndex);
+        best_time  = timesteps[i-1];
+      }
+    }
 
-  void DynamicTopologyObserver::set_cumulative_topology_modification(unsigned int type)
-  {
-    m_cumulativeTopologyModification = type;
+    changeSet->close_change_set(csIndex);
   }
+}
 
-  unsigned int DynamicTopologyObserver::get_cumulative_topology_modification() const
-  {
-    return m_cumulativeTopologyModification;
+bool file_exists(const Ioss::ParallelUtils &util,
+                 const std::string &filename,
+                 const std::string &db_type,
+                 Ioss::DatabaseUsage db_usage)
+{
+  bool exists = false;
+  int par_size = util.parallel_size();
+  int par_rank = util.parallel_rank();
+  bool is_parallel = par_size > 1;
+  std::string full_filename = filename;
+  if (is_parallel &&
+      (db_type == "exodusII" || db_type == "cgns") &&
+      (db_usage != Ioss::WRITE_HISTORY)) {
+    full_filename = Ioss::Utils::decode_filename(filename, par_rank, par_size);
   }
 
-  unsigned int DynamicTopologyObserver::get_topology_modification() const
-  {
-    return m_topologyModification;
+  if (!is_parallel || par_rank == 0) {
+    // Now, see if this file exists...
+    // Don't want to do a system call on all processors since it can take minutes
+    // on some of the larger machines, filesystems, and processor counts...
+    Ioss::FileInfo file = Ioss::FileInfo(full_filename);
+    exists = file.exists();
   }
 
-  void DynamicTopologyObserver::set_topology_modification_nl(unsigned int type)
-  {
-    m_topologyModification |= type;
-    m_cumulativeTopologyModification |= type;
+  if (is_parallel) {
```

Comment:
Is it better to use the `Ioss::FileInfo::parallel_exists` ?

Reply:
Yeah, the other version of `file_exists` makes use of it so I can convert this also